### ISSUE TITLE
Ubuntu-latest workflows will use Ubuntu-22.04

### ISSUE THREAD

**Issue by ddobranic** [actions/runner-images#6399]
### Breaking changes

Ubuntu 22.04 is ready to be the default version for the "ubuntu-latest" label in GitHub Actions and Azure DevOps.

### Target date

This change will be rolled out over a period of several weeks beginning in October, 3. We plan to complete the migration by December, 15th, 2022.

### The motivation for the changes

GitHub Actions and Azure DevOps have supported Ubuntu 22.04 in preview mode since May 2022, and starting from August 2022 Ubuntu 22.04 is generally available for all customers. We have monitored customer feedback to improve the Ubuntu 22.04 image stability and now we are ready to set it as the latest.

### Possible impact

The Ubuntu 20.04 image has a different set of software than Ubuntu 22.04. The most significant changes are listed in the table below:

|Tool name|Ubuntu 20.04|Ubuntu 22.04|Notes|
|-|:-:|-|-|
|Ruby toolcache | 2.5.9 <br>2.6.10 <br>2.7.6 <br>3.0.4 <br>3.1.2 | 3.* (latest) | Only latest available version will be cached on start.|
|Ruby | 2.7.0 | 3.* (latest) | The latest available version for each operating system provided in the APT repository is installed.|
|Clang <br>Clang-format <br>Clang-tidy | 10.0.0 <br>11.0.0 <br>12.0.0 | 12.* <br>13.* <br>14.* | Three latest available versions for each operating system provided in the APT repository are installed.|
|GCC <br>GFortran | 9.4.0 <br>10.3.0<br>- | 9.* <br>10.* <br>11.* | Three latest available versions for each operating system provided in the APT repository are installed.|
|Python | 3.8.10 | 3.10.6 | Fresh updates for Python and tools available for Ubuntu 22 in APT repository.|
|PHP | 7.4.32 <br>8.0.24 <br>8.1.11 | 8.* (latest) | Only one latest officially supported PHP version will be supported. See more [here]( https://github.com/actions/runner-images/issues/6331).|
|.NET Core SDK | 3.1.120 <br>3.1.202 <br>3.1.302 <br>3.1.423 <br>5.0.104 <br>5.0.214 <br>5.0.303 <br>5.0.408 <br>6.0.401 | 6.0.401 | Only the latest available version will be installed on start.|
|Node.js | 12.22.12<br>14.20.1<br>16.18.0<br>18.12.0 | -<br>14.20.1<br>16.18.0<br>18.12.0 | EOL version had been deprecated. If your use-case requires using any of these versions, consider using tasks to install Node.js on-flight: <br>- [actions/setup-node](https://github.com/actions/setup-node) (GitHub Actions) <br>- [Node.js Tool Installer](https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/tool/node-js?view=azure-devops) (Azure DevOps)|
|PyPy | 2.7.18 [PyPy 7.3.9] <br>3.6.12 [PyPy 7.3.3] <br>3.7.13 [PyPy 7.3.9] <br>3.8.13 [PyPy 7.3.9] <br>3.9.12 [PyPy 7.3.9] | 3.7.13 [PyPy 7.3.9] <br>3.8.13 [PyPy 7.3.9] <br>3.9.12 [PyPy 7.3.9] | Deprecated all the versions besides the most recent ones.|
|Python toolcache | 2.7.18 <br>3.6.15 <br>3.7.14 <br>3.8.14 <br>3.9.14 <br>3.10.7 | 3.7.14 <br>3.8.14 <br>3.9.14 <br>3.10.7 | On GitHub Actions, [actions/setup-python](https://github.com/actions/setup-python) can install any version on-flight so this change doesn't impact users.|

### Platforms affected

- [X] Azure DevOps
- [X] GitHub Actions

### Runner images affected

- [ ] Ubuntu 18.04
- [x] Ubuntu 20.04
- [x] Ubuntu 22.04
- [ ] macOS 10.15
- [ ] macOS 11
- [ ] macOS 12
- [ ] Windows Server 2019
- [ ] Windows Server 2022

### Mitigation ways

If you see any issues with your workflows during this transition period:

-	File an issue in this repository
-	Switch back to Ubuntu 20 by specifying the “Ubuntu-20.04" label. We will continue to support Ubuntu 20.

For more information you can take a look at the previous announcement [here](https://github.com/actions/runner-images/issues/5490).

**Comment by ddobranic** [comment]
@Kichura yes, it was a mischeck. thank you!

**Comment by joeyparrish** [comment]
Does the ubuntu-22.04 install Firefox as a snap package, or directly, or not at all?  I'm seeing test failures with that image, and worry that I'll need to either add a Firefox install to my workflow or maybe pin to ubuntu-20.04 for a while.

**Comment by Kichura** [comment]
20.04 LTS has actually shipped firefox as a snapd feature in it's desktop variant whereas the server variant does not have it at all - but can be looked into.

**Comment by joeyparrish** [comment]
We're using ubuntu-latest images for Firefox testing today.  I tried ubuntu-22.04 in a PR to see what would happen, and our workflow fails to detect Firefox at all.  So it looks like we'll have to either pin to ubuntu-20.04 or install Firefox explicitly to avoid issues.

**Comment by aravindavk** [comment]
In one of our repos, Ubuntu 22.04 based tests are failing. I am yet to figure out the issue. But the same tests working with 20.04.

https://github.com/kadalu-tech/pkgs/pull/2

```
The following packages have unmet dependencies:
 libunwind-14-dev : Breaks: libunwind-dev but 1.3.2-2build2 is to be installed
E: Error, pkgProblemResolver::Resolve generated breaks, this may be caused by held packages.
Error: Process completed with exit code 100.
```

**Comment by erik-bershel** [comment]
Hello, @aravindavk! 
[Here](https://bugs.launchpad.net/ubuntu/+source/google-glog/+bug/1991919) is an explanation why this is happening with your build. I added a [comment](https://github.com/kadalu-tech/pkgs/pull/2/files#r1001042597) explaining how to fix the situation.

**Comment by aravindavk** [comment]
Thanks @erik-bershel

**Comment by zyv** [comment]
@ddobranic it seems that Ubuntu Mozilla Team still provides official non-snap ESR packages of Firefox via PPA:

https://launchpad.net/~mozillateam/+archive/ubuntu/ppa?field.series_filter=jammy

Maybe you could coordinate with them, not having any Firefox at all in CI images is quite... limiting.

**Comment by Josh-Osmanski** [comment]
Is there any updates on Firefox not being included in ubuntu-22.04?

**Comment by ddobranic** [comment]
@Josh-Osmanski yes, we are planning to add Firefox. thank you!

**Comment by mgol** [comment]
@ddobranic The original post mentioned a migration happening automatically between October, 3 & December, 1. We're in the middle of this period. If it's indeed happening, auto-migrated repositories doing Firefox testing may start seeing breakages since Firefox has not been added to the image yet. What's the plan around that?

**Comment by MikeMcC399** [comment]
@ddobranic 

Could you update the Node.js entry in the table ^^?

The Node.js Active LTS version 18.12.0 is now cached in both [ubuntu-20.04](https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2004-Readme.md) and [ubuntu-22.04](https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2204-Readme.md). The 16.x version has also been updated.

I think it should look like this:

Tool name | Ubuntu<br>20.04 | Ubuntu<br>22.04
--- | --- | ---
Node.js | 12.22.12<br>14.20.1<br>16.18.0<br>18.12.0 | -<br>14.20.1<br>16.18.0<br>18.12.0

**Comment by ddobranic** [comment]
@MikeMcC399 thank you! updated.

**Comment by joaoantoniocardoso** [comment]
> Hello, @aravindavk! [Here](https://bugs.launchpad.net/ubuntu/+source/google-glog/+bug/1991919) is an explanation why this is happening with your build. I added a [comment](https://github.com/kadalu-tech/pkgs/pull/2/files#r1001042597) explaining how to fix the situation.

I was having the same problem, but in my case, on top of what you have suggested, I had to remove both `libunwind-14-dev` and `libunwind-14`, and then proceed as usual with the package installing. Just to mention, the dependency that was requiring `libunwind-14` was [libgstreamer1.0-dev](https://packages.debian.org/sid/libgstreamer1.0-dev). 

Thanks!

**Comment by thiell** [comment]
The table in https://github.com/actions/runner-images/blob/main/README.md still shows ubuntu-latest as ubuntu-20.04. This is confusing and should probably be fixed, thanks!

**Comment by mikhailkoliada** [comment]
@thiell the migration happens gradually, not all current ubuntu-latest resources are using ubuntu-22.04 yet. Once it is done the docs will be updated.

**Comment by chuma9615** [comment]
Hey @ddobranic

It seems that ubuntu-22.04 does not have MongoDB installed, Is there any plan to add it as a default package?

**Comment by CameronProbert** [comment]
@ddobranic Is there any update on adding Firefox as a browser to Ubuntu 22.04?

**Comment by mikhailkoliada** [comment]
@chuma9615 there is still no official ubuntu 22.04 release as far as I can see, only preview

**Comment by Bluefinger** [comment]
This updated image broke my CI WASM tests which make use of Firefox. I know this is incoming, but it's been a while since it weas stated that this is planned but things are still breaking. Is there any update on this?

**Comment by mikhailkoliada** [comment]
@Bluefinger ub22.04 with firefox will be released this week, it is already in main

**Comment by Bluefinger** [comment]
Awesome! Good to know and thanks for the update

**Comment by MikeMcC399** [comment]
@ddobranic 

Are you still planning to complete the migration of `Ubuntu-latest` to use `Ubuntu-22.04` by Dec 1, 2022? The repos which I am involved in are still using 20.04 when they run with `runs-on: ubuntu-latest` specified in the GitHub workflow.

**Comment by mikhailkoliada** [comment]
@MikeMcC399 indeed, updated :)

**Comment by MikeMcC399** [comment]
@mikhailkoliada 

I see that you changed the completion date for the migration to March 1 (2023 assumed). That is quite a significant change of 3 months from the previous Dec 1, 2022. Thank you for the update!

**Comment by 2bndy5** [comment]
I just had a [workflow that uses `ubuntu-latest`](https://github.com/cpp-linter/cpp-linter/actions/runs/3586939546) end up using ubuntu-22.04 today. In my time zone, December 1 is tomorrow, but whatver. I'm just confused when this is rolling out given that the issue is still open and the [previous comment about postponed migration](https://github.com/actions/runner-images/issues/6399#issuecomment-1331936950).

**Comment by mikhailkoliada** [comment]
@2bndy5 it is not postponed, but rather stretched, i.e. it is in progress and `ubuntu-latest` will get updates to `ubuntu-22.04`

**Comment by 2bndy5** [comment]
After I posted, I realized the deadline was a end-of-migration date, not a hard overnight change. For now, I'm using `ubuntu-22.04` to keep the workflow consistent during migration. Thanks for responding!

**Comment by MikeMcC399** [comment]
@ddobranic 

The [GitHub Changelog](https://github.blog/changelog/2022-12-01-github-actions-larger-runners-using-ubuntu-latest-label-will-now-use-ubuntu-22-04/) from Dec 1, 2022 says:

"Larger runner workflows using the ubuntu-latest runner label will soon run on [Ubuntu-22.04](https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2204-README.md).

Ubuntu-22.04 is now the [default version](https://github.blog/changelog/2022-11-09-github-actions-ubuntu-latest-workflows-will-use-ubuntu-22-04/) for the ubuntu-latest label for GitHub Actions standard runners workflows. Larger runners will now use the Ubuntu-22.04 as the -latest version starting 15 December 2022."

How is it possible to determine if a runner belongs to the category "larger runners"? What is the metric for this evaluation?

**Comment by mikhailkoliada** [comment]
@MikeMcC399 [large runners](https://docs.github.com/en/actions/using-github-hosted-runners/using-larger-runners) are paid ones.

**Comment by MikeMcC399** [comment]
@ddobranic 

> [large runners](https://docs.github.com/en/actions/using-github-hosted-runners/using-larger-runners) are paid ones.

Thank you so much for the explanation!

Can I suggest that the link to [large runners](https://docs.github.com/en/actions/using-github-hosted-runners/using-larger-runners) be added to the [GitHub Changelog](https://github.blog/changelog/2022-12-01-github-actions-larger-runners-using-ubuntu-latest-label-will-now-use-ubuntu-22-04/) from Dec 1, 2022? I didn't realise that the term "larger runners" in the blog had a special meaning.

**Comment by flcdrg** [comment]
The inclusion of only .NET 6 and 7 on 22.04 contradicts the statement in the top-level README

> 2 latest LTS versions and 1 latest version. For each feature version only latest patch is installed

According to that, 3.1 should also be included, being the second most recent LTS version

**Comment by mikhailkoliada** [comment]
@kynan python 3.6.x has been EOL since December '21, so there is no plans to add it to 22.04

**Comment by cacti77** [comment]
FYI Our C++ CI tests started failing on Linux in Azure DevOps from 1 Dec. when `ubuntu-latest` switched from 20.04 to 22.04. This failure happened whenever the test executables were started:

`symbol lookup error: /home/vsts/work/1/s/bin/os/linux_x64/test/working/../../../../../build/os/linux_x64/packages/intelmkl.devel.linux-x64.2019.5.281/lib/native/linux-x64/libmkl_intel_thread.so: undefined symbol: omp_get_num_procs`

A colleague fixed this by adding this to the pipeline yml:
`sudo apt install --no-upgrade libomp-dev`

Will we have to do this from now on? Or will you be adding libomp-dev to the ubuntu-latest (22.04) image?

**Comment by mechie** [comment]
FYI, the error message showing up in AzDO builds is misspelled (unless the underlying name *is* changing to ubntu):
```
##[warning]ubuntu-latest pipelines will use ubntu-22.04 soon. For more details, see https://github.com/actions/runner-images/issues/6399
```

**Comment by mikhailkoliada** [comment]
@mechie it depends on a pipeline type, azdo still supports `ubuntu-22.04`, say, in `vmImage`

**Comment by taspeotis** [comment]
> @mechie it depends on a pipeline type, azdo still supports `ubuntu-22.04`, say, in `vmImage`

What? He is telling you there is a typo in the warning and you are saying yes, there is a pipeline type that will accept the typo?

> ##[warning]ubuntu-latest pipelines will use **_ubntu-22.04_** soon. For more details, see https://github.com/actions/runner-images/issues/6399

I really wish Azure DevOps did not warn on this sort of stuff, it should be info: the ubuntu-latest tag is for running the latest version of Ubuntu, why are you warning me that I am getting the latest version of Ubuntu? That's what ubuntu-latest means, that's why I use it.

**Comment by taspeotis** [comment]
Weirdly it gets it right sometimes so hopefully that is a new agent image being rolled out...

![image](https://user-images.githubusercontent.com/1189399/206590301-fa8e7754-bd97-4bc7-92c3-57eed87e568b.png)

**Comment by richb-hanover** [comment]
The Original Post says:

> **Target date**
>
> This change will be rolled out over a period of several weeks beginning in October, 3. We plan to complete the migration by March, 1st, 2023.

This seems more like a several MONTH rollout. Will we endure several months of warnings that `ubuntu-latest workflows will use ubuntu-22.04 soon.`? (That's not "soon".)

I apologize for being grumpy, but unnecessary warnings make me unhappy.

**Comment by ryenus** [comment]
> **Update:** Why am I unhappy? I have 40 occurrences of `...-latest` in my small project. I'm not aware of a facility in GHA that would allow me to specify a variable that contains `ubuntu-20.04` (or `ubuntu-22.04`) and use it globally. So I'm stuck either living with the warnings or globally editing that string a load of times...

It would be much less annoying if Azure Pipelines group all the issues with the same error.

**Comment by mikeblakeuk** [comment]
"Only the latest available version will be installed on start.". so version 7 only?

**Comment by erik-bershel** [comment]
@mikeblakeuk it is already more than one: https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2204-Readme.md#net-core-sdk

**Comment by FeroxFoxxo** [comment]
> .NET Core SDK : "Only the latest available version will be installed on start.". so version 7 only?

It states that NET 6 will be supported, 6.0.401 in particular, which makes sense given that 7 is not a long-term release.

**Comment by martinlicht** [comment]
One of my actions fails on 22.04 because a C header cannot be found (`#include <omp.h>`). However, it still works with 20.04, no code changed.

What's the best way to investigate?

**Comment by taspeotis** [comment]
Google it? `omp.h not found ubuntu` yields [this](https://discourse.llvm.org/t/omp-h-not-found-after-fresh-install-with-apt-llvm-org-llvm-sh-wsl2-ubuntu-20-04/4549/3) as the second result.

> This is now also the case on Ubuntu 22.04 (which using `clang-14`). More generic install is
> `sudo apt install libomp-dev`

**Comment by couchand** [comment]
Not mentioned in the table above, but Ubuntu 22.04 ships with OpenSSL 3 by default.  I thought that pinning to 20.04 should solve it.

```diff
-runs-on: ubuntu-latest
+runs-on: ubuntu-20.04
```

But that wasn't a drop-in replacement, I immediately hit a build failure due to glibc: `error: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.33' not found`

I did not dig into this any further, and instead I will just be upgrading my systems to use OpenSSL 3.

**Comment by Vnmstheegreat** [comment]
[![No Response](https://github.com/TencentARC/GFPGAN/actions/workflows/no-response.yml/badge.svg?branch=master&event=release)](https://github.com/TencentARC/GFPGAN/actions/workflows/no-response.yml)

**Comment by virtualaidev** [comment]
Running docker-compose runs into a permission issue.

```
[2326] Failed to execute script docker-compose
Traceback (most recent call last):
  File "docker-compose", line 3, in <module>
  File "compose/cli/main.py", line 81, in main
  File "compose/cli/main.py", line 203, in perform_command
  File "compose/metrics/decorator.py", line 18, in wrapper
  File "compose/cli/main.py", line 970, in run
  File "compose/cli/main.py", line 1460, in run_one_off_container
  File "compose/cli/main.py", line 1586, in call_docker
  File "subprocess.py", line 339, in call
  File "subprocess.py", line 800, in __init__
  File "subprocess.py", line 1551, in _execute_child
PermissionError: [Errno 13] Permission denied: '/opt/hostedtoolcache/docker-stable/17.9.0-ce/x64/docker'
```

**Comment by mikhailkoliada** [comment]
ubuntu-latest has finally been switched to ubuntu 22.04 for both GitHub Actions and Azure DevOps!

**Comment by kamnyborg** [comment]
@sjd2021 it's what is installed by default, you can just add a UseDotNet-task (ex for DevOps) to install dotnet7 if you'd like

**Comment by spygi** [comment]
Is there any update on the version of PowerShell? Just double checking whether 7.2.7 is still the version [as indicated here](https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2204-Readme.md#powershell-tools).

**Comment by Qq87mica** [comment]
[payments_due_2022-12-17T21_44_22.814Z.csv](https://github.com/actions/runner-images/files/10298911/payments_due_2022-12-17T21_44_22.814Z.csv)

**Comment by jsoneaday** [comment]
What about gnome?
